src/mscorlib/src/System/Array.cs

```diff
@@ -2502,45 +2501,54 @@ sealed class SZArrayHelper {
         // This is a normal generic Enumerator for SZ arrays. It doesn't have any of the "this" stuff
         // that SZArrayHelper does.
         //
-        [Serializable] private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T> {
+        [Serializable]
+        private sealed class SZGenericArrayEnumerator<T> : IEnumerator<T>
+        {
             private T[] _array;
             private int _index;
-            private int _endIndex; // cache array length, since it's a little slow.
 
-            // Passing -1 for endIndex so that MoveNext always returns false without mutating _index
-            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(null, -1);
+            internal static readonly SZGenericArrayEnumerator<T> Empty = new SZGenericArrayEnumerator<T>(Array.Empty<T>());
```

Comment:
Here, we could check if `T` is a common type like `byte` or `char` (or `int`?) and use `Array.Empty<T>` in such cases, otherwise allocate an empty `T[0]`. The JIT should elide the type checks. Probably overkill.